### PR TITLE
feat(cmf): get the http response to make something with it

### PR DESCRIPTION
**Pull request by jare-talend**
**Please check if the PR fulfills these requirements**

- [X] The commit(s) message(s) follows our
  [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)
Currently, when a error http triggers, we lost the response object. 

**What is the new behavior?**
The response object is returned. We can get the body from the server, get a custom error provides by the server and show it for the user.
The object response can be used when you pass a function into an onError.
On the @@HTTP/ERRORS, now, you can access to the body and custom the error message. We need to parse the return into the middleware to avoid a asynchronous function into the reducers.

```
// I can handle my callback to make something when I have a HTTP error
dispatch(actions.http.post('url',
   body,
{
   onError: (error) => {
console.log(error.stack.response.messageObject)
},
}
));

// i can use the parsed body into my action '@@HTTP/ERRORS'
export default function notifications(state, action) {
	switch (action.type) {
	case '@@HTTP/ERRORS':
		return pushNotification(state, {
			type: 'error',
			message: action.error.stack.messageObject
		});
}}
```

**Does this PR introduce a breaking change?**

- [ ] Yes
- [X] No

If this PR contains a breaking change, please describe the impact and migration
path for existing applications: ...

**Other information**:

### CODE REVIEW

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://212.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://212.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://212.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://212.talend.surge.sh)

File: packages/cmf/src/middlewares/http/middleware.js [hunk 1]
```diff
@@ -12,7 +12,7 @@ export const DEFAULT_HTTP_HEADERS = {
 };
 
 export function isHTTPRequest(action) {
-	return action.type in HTTP_METHODS;
+	return action && action.type in HTTP_METHODS;
```
Comment (jmfrancois) [inline packages/cmf/src/middlewares/http/middleware.js]: type is required attribut in redux

File: packages/cmf/src/middlewares/http/middleware.js [hunk 2]
```diff
@@ -86,6 +86,7 @@ export function HTTPError(response) {
 	this.name = `HTTP ${response.status}`;
 	this.message = response.statusText;
 	this.stack = {
+		response,
```
Comment (jmfrancois) [inline packages/cmf/src/middlewares/http/middleware.js]: response is not a plain object. It's a full of promise.
I would prefer a serialization of it (extract the body and add the info to the stack)  see next comment

File: packages/cmf/src/middlewares/http/middleware.js [hunk 3]
```diff
@@ -136,7 +137,16 @@ export const httpMiddleware = ({ dispatch }) => next => (action) => {
...
 			},
 		}, action);
-		dispatch(httpError(newAction.error));
+
+		// clone the response object else the next call to text or json
+		// triggers an exception Already use
+		newAction.error.stack.response.clone().text().then((response) => {
+			try {
+				newAction.error.stack.messageBody = JSON.parse(response);
+			} finally {
```
Comment (jmfrancois) [inline packages/cmf/src/middlewares/http/middleware.js]:
```
stack.response = response;
stack.messageObject = JSON.parse(response);
```

File: packages/cmf/src/middlewares/http/middleware.js (hunk 1 above)
Comment (jare-talend) [inline packages/cmf/src/middlewares/http/middleware.js]: If the user provide a custom method on the onError and this method doesn't provide a type, this trigger errors on CMF. I can add a check on this.

File: packages/cmf/src/middlewares/http/middleware.js (hunk 2 above)
Comment (jare-talend) [inline packages/cmf/src/middlewares/http/middleware.js]: I need this to handle this in the onHTTPError. If the text() sucess, it crushes it with the serialization.

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://212.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://212.talend.surge.sh)

File: packages/cmf/__tests__/middlewares/http.test.js [hunk 4]
```diff
@@ -176,12 +179,141 @@ describe('CMF http middleware', () => {
...
+		const store = {
+			dispatch: jest.fn(),
+		};
+		const next = jest.fn();
+		const action = {
+			type: HTTP_METHODS.POST,
+			body: { label: 'great test' },
+			onSend: 'CALL_ME_BACK on send',
+			onResponse: 'CALL_ME_BACK on response',
+			onError: () => {},
```
Comment (jmfrancois) [inline packages/cmf/__tests__/middlewares/http.test.js]: no empty action creator please

File: packages/cmf/src/middlewares/http/middleware.js [hunk 5]
```diff
@@ -136,10 +137,23 @@ export const httpMiddleware = ({ dispatch }) => next => (action) => {
...
+			try {
+				newAction.error.stack.response = response;
+				newAction.error.stack.messageObject = JSON.parse(response);
+			} finally {
+				dispatch(httpError(newAction.error));
+				if (newAction.onError) {
+					const onErrorAction = onError(newAction, newAction.error);
+					if (onErrorAction) {
+						dispatch(onError(newAction, newAction.error));
+					}
```
Comment (jmfrancois) [inline packages/cmf/src/middlewares/http/middleware.js]: if the function exists it's an action creator so it must return an action object.
here you fail silently. please remove this if and just call the dispatch on it
dispatch(onError(newAction, newAction.error));

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://212.talend.surge.sh)

**Pull request merged**